File: linux/advanced-prog-in-unix/chap12/liftoff.c
/*
 * liftoff.c
 * Emulator rockets lifting off using thread.
 * The counting down before lanuching is by thread.
 */
#include <pthread.h>
#include <stdio.h>
#include <unistd.h>

static void *liftoff( void *arg ) {
  int a = *(int *) arg;
  while ( a ) {
    printf( " %d ", a-- );
  }
  printf( "Lift off...\n" );

  return (void *) 0;
}

int main( void ) {
  pthread_t tid;
  for ( int i = 1; i <= 10; i++ ) {
    pthread_create( &tid, NULL, liftoff, (void *) &i );
  }
  printf( "Waiting for lift off:\n" );
  sleep( 10 );
  return 0;
}

File: linux/advanced-prog-in-unix/chap12/LiftOff.java
public class LiftOff implements Runnable {
  private int count;
  public LiftOff( int count ) {
    this.count = count;
  }

  public void run() {
    try {
      Thread.sleep( 1000 );
    } catch ( InterruptedException e ) {
      e.printStackTrace();
    }
    while ( count > 0 ) {
      System.out.print( count-- + " " );
    }
    System.out.println( "Lift off" );
  }

  public static void  main( String[] args ) {
    for ( int i = 0; i < 10; i++ ) {
      new LiftOff( 10 ).run();
    }

    System.out.println( "Waiting for lift off" );
    try {
      Thread.sleep( 5000 );
    } catch ( InterruptedException e ) {
      e.printStackTrace();
    }
  }
}
